### PR TITLE
Limechain/16xx

### LINKED ISSUE #456: Users SNBAT perform actions that remove all SortedOracles reports
### Expected Behavior

reports.length > 0 always

### Current Behavior

If all reports expire or all oracles are removed, reports.length can equal 0

### LINKED ISSUE #1379: UsingRegistry should expose getters for all registered contracts
### Expected Behavior

getters for all registered contracts exist, assuming this does not increase the bytecode size of contracts that only use a subset of those getters

### Current Behavior

They do not

### LINKED ISSUE #1601: Exchange fallback function should not be payable
### Expected Behavior

No fallback function, particularly not one that's payable. This would allow users to send the exchange Celo Gold without it being withdrawable.

### Current Behavior

Payable fallback function. Possibly needed for testing. If this is the case, we can create a separate `ExchangeTest.sol` that inherits from `Exchange` that adds a payable fallback function.

### LINKED ISSUE #1602: UsingRegistry should check that registry address is not zero in setRegistry
### Expected Behavior

Please describe the behavior you are expecting

### Current Behavior

What is the current behavior?

### LINKED ISSUE #1605: MultiSig should not be able to be owned by the null address
### Expected Behavior
`addOwner` and `replaceOwner` should use the `notNull` modifier, tests that cover this.

### Current Behavior

Could accidentally add null address as owner

### LINKED ISSUE #1606: Integer overflow vulnerability in Escrow
### Expected Behavior

Safemath everywhere, always.

### Current Behavior

`now >= (payment.timestamp + payment.expirySeconds)`

### LINKED ISSUE #1841: Set a minimum gas price minimum
### Expected Behavior

A parameter (set in migrations config as a default of 1 GWei ) to set a floor on the gas price minimum so that even with empty blocks GPM cannot drift below that value.  

Then verify that value is shown correctly via ethstats.

### Current Behavior

GPM can fall to 4

### PR DESCRIPTION
**Pull request by BatiGencho**
### Description

Small fixes to  few issues described in the sections below.

### Related issues

- Fixes #1601 
- Fixes #1602 
- Fixes #1605 
- Fixes #1606 
- Fixes #1841 
- Fixes #1379 
- Fixes #456

### CODE REVIEW

File: packages/protocol/contracts/common/UsingRegistry.sol [hunk 1]
```diff
@@ -54,6 +54,7 @@ contract UsingRegistry is Ownable {
    * @param registryAddress The address of a registry contract for routing to other contracts.
    */
   function setRegistry(address registryAddress) public onlyOwner {
+    require(registryAddress != address(0), "attempt to set the reserved 0x0 address to registry");
```
Comment (asaj) [inline packages/protocol/contracts/common/UsingRegistry.sol]: nit: "Cannot register the null address"

File: packages/protocol/test/common/multisig.ts.old [hunk 2]
```diff
@@ -133,6 +133,10 @@ contract('MultiSig', (accounts: any) => {
     it('should not allow an external account to add an owner', async () => {
       await assertRevert(multiSig.addOwner(accounts[3]))
     })
+
+    it('should not allow a new owner - the 0x address to be added', async () => {
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts.old]: Pleaseename this file to `multisig.ts`

File: packages/protocol/test/common/multisig.ts.old [hunk 3]
```diff
@@ -133,6 +133,10 @@ contract('MultiSig', (accounts: any) => {
...
+      await assertRevert(multiSig.contract.addOwner.getData(NULL_ADDRESS))
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts.old]: Why not `multisig.addOwner(NULL_ADDRESS)`?

File: packages/protocol/test/common/multisig.ts.old [hunk 4]
```diff
@@ -167,6 +171,14 @@ contract('MultiSig', (accounts: any) => {
     it('should not allow an external account to replace an owner', async () => {
       await assertRevert(multiSig.replaceOwner(accounts[1], accounts[2]))
     })
+
+    it('should not allow a new owner - the 0x address to be replaced by an old owner', async () => {
+      await assertRevert(multiSig.contract.replaceOwner.getData(accounts[1], NULL_ADDRESS))
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts.old]: Ditto, why use `.contract` and `.getData` here?

File: packages/protocol/test/common/multisig.ts.old [hunk 5]
```diff
@@ -167,6 +171,14 @@ contract('MultiSig', (accounts: any) => {
...
+    })
+
+    it('should not allow the 0x address to call the replace owner function', async () => {
+      await assertRevert(multiSig.contract.replaceOwner.getData(NULL_ADDRESS, accounts[2]))
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts.old]: Ditto

File: packages/protocol/test/common/multisig.ts.old (hunk 2 above)
Comment (asaj) [inline packages/protocol/test/common/multisig.ts.old]: For consistency with our other tests: "it should not allow adding the null address"

File: packages/protocol/test/common/multisig.ts.old [hunk 6]
```diff
@@ -167,6 +171,14 @@ contract('MultiSig', (accounts: any) => {
     it('should not allow an external account to replace an owner', async () => {
       await assertRevert(multiSig.replaceOwner(accounts[1], accounts[2]))
     })
+
+    it('should not allow a new owner - the 0x address to be replaced by an old owner', async () => {
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts.old]: "should not allow an owner to be replaced by the null address"

File: packages/protocol/test/common/multisig.ts.old [hunk 7]
```diff
@@ -167,6 +171,14 @@ contract('MultiSig', (accounts: any) => {
...
+      await assertRevert(multiSig.contract.replaceOwner.getData(accounts[1], NULL_ADDRESS))
+    })
+
+    it('should not allow the 0x address to call the replace owner function', async () => {
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts.old]: Are you sure that this is actually what's being tested?

File: packages/protocol/contracts/common/MultiSig.sol [hunk 8]
```diff
@@ -155,6 +155,8 @@ contract MultiSig is Initializable {
     external
     onlyWallet
     ownerExists(owner)
+    notNull(owner)
```
Comment (asaj) [inline packages/protocol/contracts/common/MultiSig.sol]: Is this necessary?

**Comment by asaj** [comment]
Please include any additional tickets you've addressed in the PR description

File: packages/protocol/contracts/common/GasPriceMinimum.sol [hunk 9]
```diff
@@ -15,10 +15,11 @@ contract GasPriceMinimum is Ownable, Initializable, UsingRegistry {
   using SafeMath for uint256;
 
   event TargetDensitySet(uint256 targetDensity);
-
+  event GasPriceMinimumThresholdSet(uint256 gasPriceMinimumThreshold);
```
Comment (asaj) [inline packages/protocol/contracts/common/GasPriceMinimum.sol]: nit: Consider `GasPriceMinimumFloor` instead, here and everywhere

File: packages/protocol/contracts/common/GasPriceMinimum.sol [hunk 10]
```diff
@@ -34,12 +35,14 @@ contract GasPriceMinimum is Ownable, Initializable, UsingRegistry {
...
     address _registryAddress,
     uint256 initialGas,
+    uint256 _gasPriceMinimumThreshold,
     uint256 _targetDensity,
     uint256 _adjustmentSpeed
   ) external initializer {
     _transferOwnership(msg.sender);
     setRegistry(_registryAddress);
     gasPriceMinimum = initialGas;
+    gasPriceMinimumThreshold = _gasPriceMinimumThreshold;
```
Comment (asaj) [inline packages/protocol/contracts/common/GasPriceMinimum.sol]: Please use `setGasPriceMinimumThreshold` for this

File: packages/protocol/contracts/common/UsingRegistry.sol [hunk 11]
```diff
@@ -62,10 +67,22 @@ contract UsingRegistry is Ownable {
...
+
   function getElection() internal view returns (IElection) {
     return IElection(registry.getAddressForOrDie(ELECTION_REGISTRY_ID));
   }
 
+  function getExchange() internal view returns (IExchange) {
+    return IExchange(registry.getAddressForOrDie(EXCHANGE_REGISTRY_ID));
+  }
+
+  function getGasCurrencyWhitelistRegistry() internal view returns (IFeeCurrencyWhitelist) {
```
Comment (asaj) [inline packages/protocol/contracts/common/UsingRegistry.sol]: s/Gas/Fee?

File: packages/protocol/contracts/common/UsingRegistry.sol [hunk 12]
```diff
@@ -78,6 +95,10 @@ contract UsingRegistry is Ownable {
     return ILockedGold(registry.getAddressForOrDie(LOCKED_GOLD_REGISTRY_ID));
   }
 
+  function getReserve() internal view returns (IReserve) {
```
Comment (asaj) [inline packages/protocol/contracts/common/UsingRegistry.sol]: Please sort alphabetically

File: packages/protocol/migrationsConfig.js [hunk 13]
```diff
@@ -50,7 +50,8 @@ const DefaultConfig = {
     minimumReports: 1,
   },
   gasPriceMinimum: {
-    initialMinimum: 10000,
+    initialMinimum: 1000000000,
```
Comment (asaj) [inline packages/protocol/migrationsConfig.js]: Rather than pass an initial minimum, let's set the initial value to the floor. So we can remove this from the migration config, and the constructor, tests, etc

File: packages/protocol/scripts/bash/ganache.sh [hunk 14]
```diff
@@ -9,5 +9,5 @@ yarn run ganache-cli \
   --gasPrice 0 \
   --networkId 1101 \
   --gasLimit 10000000 \
-  --defaultBalanceEther 1000000 \
+  --defaultBalanceEther 200000000 \
```
Comment (asaj) [inline packages/protocol/scripts/bash/ganache.sh]: Why is this needed?

File: packages/protocol/test/common/gaspriceminimum.ts [hunk 15]
```diff
@@ -133,11 +141,47 @@ contract('GasPriceMinimum', (accounts: string[]) => {
...
+
   describe('#getUpdatedGasPriceMinimum', () => {
     describe('when the block is full', () => {
-      it('should return 25% more than the initial minimum', async () => {
+      it('should return 25% more than the initial minimum and should not be limited by the gas price minimum threshold as a whole', async () => {
+        const currentGasPriceMinimum = await gasPriceMinimum.gasPriceMinimum()
+        await gasPriceMinimum.setGasPriceMinimumThreshold(currentGasPriceMinimum)
         const actualUpdatedGasPriceMinimum = await gasPriceMinimum.getUpdatedGasPriceMinimum(1, 1)
-        const expectedUpdatedGasPriceMinimum = initialGasPriceMinimum
+        const expectedUpdatedGasPriceMinimum = new BigNumber(currentGasPriceMinimum.toString())
```
Comment (asaj) [inline packages/protocol/test/common/gaspriceminimum.ts]: nit: currentGasPriceMinimum should be a bignumber already

File: packages/protocol/test/common/gaspriceminimum.ts [hunk 16]
```diff
@@ -146,9 +190,19 @@ contract('GasPriceMinimum', (accounts: string[]) => {
...
+        const expectedCappedUpdatedGasPriceMinimum = await gasPriceMinimum.gasPriceMinimumThreshold()
+        assertEqualBN(actualUpdatedGasPriceMinimum, expectedCappedUpdatedGasPriceMinimum)
+      })
+
+      it('should return 25% less than the initial minimum, but not be limited by the gas price minimum threshold if new gas price lies above minimum', async () => {
+        const currentGasPriceMinimum = await gasPriceMinimum.gasPriceMinimum()
+        await gasPriceMinimum.setGasPriceMinimumThreshold(1)
         const actualUpdatedGasPriceMinimum = await gasPriceMinimum.getUpdatedGasPriceMinimum(0, 1)
-        const expectedUpdatedGasPriceMinimum = initialGasPriceMinimum
+        const expectedUpdatedGasPriceMinimum = new BigNumber(currentGasPriceMinimum.toString())
```
Comment (asaj) [inline packages/protocol/test/common/gaspriceminimum.ts]: ditto

File: packages/protocol/test/common/multisig.ts [hunk 17]
```diff
@@ -35,21 +79,26 @@ contract('MultiSig', (accounts: any) => {
   describe('#submitTransaction()', () => {
     let txData: string
     beforeEach(async () => {
-      // @ts-ignore contract is a property
-      txData = multiSig.contract.addOwner.getData(accounts[2])
+      txData = encodeAddOwnerFunc(web3, accounts[2])
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts]: The `multiSig` object already has the ABI stored, why not leverage that via `getData`?

If this was due to https://github.com/celo-org/celo-monorepo/pull/1982/commits/39e33b794ceea96796f91ff6b39f58f38a144460#r353409066, that may have been a miscommunication, since I was suggesting that we shouldn't use `getData` to make a contract call, and not that it wasn't a good way of encoding tx data

File: packages/protocol/test/common/multisig.ts [hunk 18]
```diff
@@ -133,18 +203,32 @@ contract('MultiSig', (accounts: any) => {
     it('should not allow an external account to add an owner', async () => {
       await assertRevert(multiSig.addOwner(accounts[3]))
     })
+
+    it('should not allow adding the null address', async () => {
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts:168]: I don't think this is testing what we'd like it to. It's failing because it's not being called by the `multisig` address, and not because it's adding the `NULL_ADDRESS`.

File: packages/protocol/test/common/multisig.ts [hunk 19]
```diff
@@ -133,18 +203,32 @@ contract('MultiSig', (accounts: any) => {
...
+      })
+      // @ts-ignore: TODO(mcortesi): fix typings
+      const txId = txEvent.args.transactionId
+
       // @ts-ignore: TODO(mcortesi): fix typings
       await multiSig.confirmTransaction(txId, { from: accounts[1] })
+
       assert.isFalse(await multiSig.isOwner(accounts[1]))
-      assert.equal(await multiSig.required(), new BigNumber(1))
+      assert.equal((await multiSig.required()).toNumber(), 1)
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts]: Use assertEqualBN here

File: packages/protocol/test/common/multisig.ts [hunk 20]
```diff
@@ -167,17 +258,29 @@ contract('MultiSig', (accounts: any) => {
     it('should not allow an external account to replace an owner', async () => {
       await assertRevert(multiSig.replaceOwner(accounts[1], accounts[2]))
     })
+
+    it('should not allow an owner to be replaced by the null address', async () => {
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts:239]: Ditto, this is failing because it's not being called by the multisig. We're not actually testing what we intend to test here

File: packages/protocol/test/common/multisig.ts [hunk 21]
```diff
@@ -167,17 +258,29 @@ contract('MultiSig', (accounts: any) => {
...
+      const txEvent = _.find(tx.logs, {
+        event: 'Confirmation',
+      })
+      // @ts-ignore: TODO(mcortesi): fix typings
+      const txId = txEvent.args.transactionId
+
       // @ts-ignore: TODO(mcortesi): fix typings
       await multiSig.confirmTransaction(txId, { from: accounts[1] })
-      assert.equal(await multiSig.required(), new BigNumber(1))
+      assert.equal((await multiSig.required()).toNumber(), 1)
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts]: assertEqualBN

File: packages/protocol/test/common/multisig.ts [hunk 22]
```diff
@@ -188,27 +291,35 @@ contract('MultiSig', (accounts: any) => {
...
+      const txEvent = _.find(tx.logs, {
+        event: 'Confirmation',
+      })
+      // @ts-ignore: TODO(mcortesi): fix typings
+      txId = txEvent.args.transactionId
     })
 
     it('should return the confirmation count', async () => {
-      assert.equal(await multiSig.getConfirmationCount(txId), new BigNumber(1))
+      assert.equal((await multiSig.getConfirmationCount(txId)).toNumber(), 1)
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts]: ditto

File: packages/protocol/test/common/multisig.ts [hunk 23]
```diff
@@ -188,27 +291,35 @@ contract('MultiSig', (accounts: any) => {
...
+      // @ts-ignore: TODO(mcortesi): fix typings
+      const txData = encodeAddOwnerFunc(web3, accounts[2])
+      await multiSig.submitTransaction(multiSig.address, 0, txData, {
+        from: accounts[0],
+      })
     })
 
     it('should return the transaction count', async () => {
-      assert.equal(await multiSig.getTransactionCount(true, true), new BigNumber(1))
+      assert.equal((await multiSig.getTransactionCount(true, true)).toNumber(), 1)
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts]: ditto

**Review by asaj — COMMENTED** [review]
Thanks for reviving the multisig tests!

File: packages/protocol/contracts/common/GasPriceMinimum.sol (hunk 9 above)
Comment (BatiGencho) [reply packages/protocol/contracts/common/GasPriceMinimum.sol]: done

File: packages/protocol/contracts/common/GasPriceMinimum.sol (hunk 10 above)
Comment (BatiGencho) [reply packages/protocol/contracts/common/GasPriceMinimum.sol]: done

File: packages/protocol/contracts/common/UsingRegistry.sol (hunk 11 above)
Comment (BatiGencho) [reply packages/protocol/contracts/common/UsingRegistry.sol]: changed

File: packages/protocol/contracts/common/UsingRegistry.sol (hunk 12 above)
Comment (BatiGencho) [reply packages/protocol/contracts/common/UsingRegistry.sol]: done

File: packages/protocol/migrationsConfig.js (hunk 13 above)
Comment (BatiGencho) [reply packages/protocol/migrationsConfig.js]: done

File: packages/protocol/scripts/bash/ganache.sh (hunk 14 above)
Comment (BatiGencho) [reply packages/protocol/scripts/bash/ganache.sh]: because the inital ether balance has been increased in all other ganache servers (required by the new migtrations config), so I thought that if one was to run it on its own and deploy himself, it would be nice to have the same ether balance s in the other scripts

File: packages/protocol/test/common/gaspriceminimum.ts (hunk 15 above)
Comment (BatiGencho) [reply packages/protocol/test/common/gaspriceminimum.ts]: changed

File: packages/protocol/test/common/gaspriceminimum.ts (hunk 16 above)
Comment (BatiGencho) [reply packages/protocol/test/common/gaspriceminimum.ts]: done

File: packages/protocol/test/common/multisig.ts (hunk 17 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts]: well, since Truffle V5 getData does has been repaced by the underlying web3 object, hence that would not work anymore. I changed it appropriately

File: packages/protocol/test/common/multisig.ts (hunk 18 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts:168]: true, neither was this well thought through in the old tests, so I changed it

File: packages/protocol/test/common/multisig.ts (hunk 19 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts]: done

File: packages/protocol/test/common/multisig.ts (hunk 21 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts]: done

File: packages/protocol/test/common/multisig.ts (hunk 23 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts]: done

File: packages/protocol/test/common/multisig.ts (hunk 22 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts]: dine

File: packages/protocol/test/common/multisig.ts (hunk 20 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts:239]: redesigned

File: packages/protocol/test/common/multisig.ts [hunk 24]
```diff
@@ -0,0 +1,382 @@
...
+      assert.isTrue(await multiSig.confirmations(txId, accounts[0]))
+      assert.equal((await multiSig.transactionCount()).toNumber(), 1)
+    })
+
+    it('should not allow an owner to submit a transaction to a null address', async () => {
+      // @ts-ignore: TODO(mcortesi): fix typings
+      await assertRevert(multiSig.submitTransaction(NULL_ADDRESS, 0, txData))
+    })
+
+    it('should not allow an owner to submit a transaction', async () => {
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts]: "a non-owner"?

File: packages/protocol/test/common/multisig.ts [hunk 25]
```diff
@@ -0,0 +1,382 @@
...
+      const txEvent = _.find(tx.logs, {
+        event: 'Confirmation',
+      })
+      const txId = txEvent.args.transactionId
+      // @ts-ignore: TODO(mcortesi): fix typings
+      await multiSig.confirmTransaction(txId, { from: accounts[1] })
+      assert.isTrue(await multiSig.isOwner(accounts[2]))
+    })
+
+    it('should not allow an external account to add an owner', async () => {
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts:163]: Should this test instead be calling `addOwner` directly rather than `submitTransaction`?

File: packages/protocol/test/common/multisig.ts [hunk 26]
```diff
@@ -0,0 +1,382 @@
...
+      const txId = txEvent.args.transactionId
+
+      // @ts-ignore: TODO(mcortesi): fix typings
+      await multiSig.confirmTransaction(txId, { from: accounts[1] })
+
+      assert.isFalse(await multiSig.isOwner(accounts[1]))
+      assertEqualBN(await multiSig.required(), 1)
+    })
+
+    it('should not allow an external account to remove an owner', async () => {
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts:209]: Ditto, should this call `removeOwner` directly?

File: packages/protocol/test/common/multisig.ts [hunk 27]
```diff
@@ -0,0 +1,382 @@
...
+      })
+      // @ts-ignore: TODO(mcortesi): fix typings
+      const txId = txEvent.args.transactionId
+      // @ts-ignore: TODO(mcortesi): fix typings
+      await multiSig.confirmTransaction(txId, { from: accounts[1] })
+      assert.isTrue(await multiSig.isOwner(accounts[2]))
+      assert.isFalse(await multiSig.isOwner(accounts[1]))
+    })
+
+    it('should not allow an external account to replace an owner', async () => {
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts:234]: Ditto

File: packages/protocol/test/common/multisig.ts [hunk 28]
```diff
@@ -0,0 +1,382 @@
...
+      // @ts-ignore: TODO(mcortesi): fix typings
+      const txEvent = _.find(tx.logs, {
+        event: 'Confirmation',
+      })
+      // @ts-ignore: TODO(mcortesi): fix typings
+      const txId = txEvent.args.transactionId
+
+      // @ts-ignore: TODO(mcortesi): fix typings
+      await multiSig.confirmTransaction(txId, { from: accounts[1] })
+      assert.equal((await multiSig.required()).toNumber(), 1)
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts]: assertEqualBN?

File: packages/protocol/test/common/multisig.ts [hunk 29]
```diff
@@ -0,0 +1,382 @@
...
+    })
+
+    it('should not allow an external account to change the requirement', async () => {
```
Comment (asaj) [inline packages/protocol/test/common/multisig.ts:276]: Ditto, should this call `changeRequirement` directly?

**Review by asaj — COMMENTED** [review]
Looks great! Just some minor comments on the multisig tests

File: packages/protocol/test/common/multisig.ts (hunk 24 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts]: fixed

File: packages/protocol/test/common/multisig.ts (hunk 25 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts:163]: changed

File: packages/protocol/test/common/multisig.ts (hunk 26 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts:209]: changed

File: packages/protocol/test/common/multisig.ts (hunk 27 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts:234]: changed

File: packages/protocol/test/common/multisig.ts (hunk 28 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts]: changed

File: packages/protocol/test/common/multisig.ts (hunk 29 above)
Comment (BatiGencho) [reply packages/protocol/test/common/multisig.ts:276]: changed

**Review by mcortesi — APPROVED** [review]
LGTM

**Review by mcortesi — APPROVED** [review]
